src/Engine/Shaders/ParticleShader.cpp

```diff
@@ -0,0 +1,273 @@
+            break;
+        }
+
+        // The reason there is a plus one is that there has to always be at least one element in RandomMs even if unused.
+        if (pc->RandomMs().size() > pc->MAX_ORIENTATIONS + 1) {
+            break;
+        }
+
+        /* Toon shading */
+        if (showToon) {
+            loadBool(getUniform("isToon"), true);
+        }
+        else {
+            loadBool(getUniform("isToon"), false);
+        }
+
+        bool r = (int)pc->RandomMs().size() != 1;
+        loadBool(getUniform("randomOrientation"), r);
+
+        /* Model matrix */
+        
+        /* Load Random Matrix Array */
+        loadMat4(getUniform("M"), pc->ModelMatrix());
+        loadMultiMat4(getUniform("randomMs"), &(pc->RandomMs()[0]), (int)pc->RandomMs().size());
+        
+        /* Load Normal matrix */
+        loadMat3(getUniform("N"), pc->NormalMatrix());
+        loadMultiMat3(getUniform("randomNs"), &(pc->RandomNs()[0]), (int)pc->RandomNs().size());
+
+
+
+        /* Bind materials */
+        loadFloat(getUniform("matAmbient"), pc->getModelTexture(0)->material.ambient);
+        loadVec3(getUniform("matDiffuse"), pc->getModelTexture(0)->material.diffuse);
+        loadVec3(getUniform("matSpecular"), pc->getModelTexture(0)->material.specular);
+        loadFloat(getUniform("shine"), pc->getModelTexture(0)->material.shineDamper);
+   
+        /* Load texture */
+        if(pc->getModelTexture(0)->texture && pc->getModelTexture(0)->texture->textureId != 0) {
+            loadBool(getUniform("usesTexture"), true);
+            loadInt(getUniform("textureImage"), pc->getModelTexture(0)->texture->textureId);
+            glActiveTexture(GL_TEXTURE0 + pc->getModelTexture(0)->texture->textureId);
+            glBindTexture(GL_TEXTURE_2D, pc->getModelTexture(0)->texture->textureId);
+        }
+        else {
+            loadBool(getUniform("usesTexture"), false);
+        }
+
+        /* Bind mesh */
+        glBindVertexArray(pc->getMesh(0)->vaoId);
+
+        /*Set Particle positions*/
+        int posPO = getAttribute("particleOffset");
+        unsigned int offsetBufId;
+        Vector<glm::vec3> *positions = pc->getParticlePositions();
+
+        /* Do all the buffer stuff */
+        glGenBuffers(1, &offsetBufId);
+        glBindBuffer(GL_ARRAY_BUFFER, offsetBufId);
+        glBufferData(GL_ARRAY_BUFFER, positions->size() * sizeof(glm::vec3), &(*positions)[0], GL_STATIC_DRAW);
+        glEnableVertexAttribArray(posPO);
+        glVertexAttribPointer(posPO, 3, GL_FLOAT, GL_FALSE, 0, (const void *)0);
+
+        /* Update particle position attribute once per instance */
+        glVertexAttribDivisor(posPO, 1);
+
+        /* Set Orientation IDs*/
+        int randOrient = getAttribute("orientationID");
+        unsigned int orientBufId;
+        Vector<int> *orientations = pc->getParticleOrientationIDs();
+
+        /* Do all the buffer stuff */
+        glGenBuffers(1, &orientBufId);
```

Comment:
Again, we shouldn't be generating a new buffer per particle per frame

Reply:
I'll fix this.

Reply:
see above